src/chunkformatter.js

```diff
@@ -0,0 +1,311 @@
+ */
+ChunkFormatter.prototype.validateNewRow = function(chunk) {
+  const row = this.row;
+  const prevRowKey = this.prevRowKey;
+  const newRowKey = Mutation.convertFromBytes(chunk.rowKey);
+  this.isError(
+    typeof row.key !== 'undefined',
+    'A new row cannot have existing state',
+    chunk
+  );
+  this.isError(
+    typeof chunk.rowKey === 'undefined' || chunk.rowKey === '',
+    'A row key must be set',
+    chunk
+  );
+  this.isError(chunk.resetRow, 'A new row cannot be reset', chunk);
+  this.isError(
+    prevRowKey === newRowKey,
+    'A commit happened but the same key followed',
+    chunk
+  );
+  this.isError(!chunk.familyName, 'A family must be set', chunk);
+  this.isError(!chunk.qualifier, 'A column qualifier must be set', chunk);
+  this.validateValueSizeAndCommitRow(chunk);
+};
+/**
+ * Validates state for rowInProgress
+ * @private
+ * @param {chunk} chunk chunk to validate
+ */
+ChunkFormatter.prototype.validateRowInProgress = function(chunk) {
+  const row = this.row;
+  this.validateResetRow(chunk);
+  const newRowKey = Mutation.convertFromBytes(chunk.rowKey);
+  this.isError(
+    chunk.rowKey && newRowKey !== row.key,
+    'A commit is required between row keys',
+    chunk
+  );
+  this.validateValueSizeAndCommitRow(chunk);
+  this.isError(
+    chunk.familyName && !chunk.qualifier,
+    'A qualifier must be specified',
+    chunk
+  );
+};
+/**
+ * Validates chunk for cellInProgress state.
+ * @private
+ * @param {chunk} chunk chunk to validate
+ */
+ChunkFormatter.prototype.validateCellInProgress = function(chunk) {
+  this.validateResetRow(chunk);
+  this.validateValueSizeAndCommitRow(chunk);
+};
+/**
+ * Moves to next state in processing.
+ * @private
+ * @param {chunk} chunk chunk in process
+ * @param {*} callback callback to call with row as and when generates
+ */
+ChunkFormatter.prototype.moveToNextState = function(chunk, callback) {
+  const row = this.row;
+  if (chunk.commitRow) {
+    callback(null, row);
+    this.commit();
+  } else {
+    if (chunk.valueSize > 0) {
+      this.state = RowStateEnum.CELL_IN_PROGRESS;
+    } else {
+      this.state = RowStateEnum.ROW_IN_PROGRESS;
+    }
+  }
+};
+/**
+ * Process chunk when in NEW_ROW state.
+ * @private
+ * @param {chunks} chunk chunk to process
+ * @param {options} options options
+ * @param {callback} callback callback to call with row as and when generates
+ */
+ChunkFormatter.prototype.newRow = function(chunk, options, callback) {
+  const newRowKey = Mutation.convertFromBytes(chunk.rowKey);
+  this.validateNewRow(chunk);
+  const row = this.row;
+  row.key = newRowKey;
+  row.data = {};
+  this.family = row.data[chunk.familyName.value] = {};
+  const qualifierName = Mutation.convertFromBytes(chunk.qualifier.value);
+  this.qualifiers = this.family[qualifierName] = [];
+  this.qualifier = {
+    value: Mutation.convertFromBytes(chunk.value, options),
+    labels: chunk.labels,
+    timestamp: chunk.timestampMicros,
+    size: chunk.valueSize,
```

Comment:
Is there any value in having size as part of the data? this.qualifier is the actual data in which user is interested in. chunk.valueSize is meant for the API to indicate when the Cell is complete. It helps API to pre-allocate a buffer to store the information and usually in bytes. Besides, everything is converted from bytes so it looses its relevance.